semantic_model_generator/admin_app/app_utils.py

```diff
@@ -0,0 +1,741 @@
+import os
+from datetime import datetime
+from io import StringIO
+
+import pandas as pd
+import streamlit as st
+
+from semantic_model_generator.data_processing.proto_utils import (
+    proto_to_yaml,
+    yaml_to_semantic_model,
+)
+from semantic_model_generator.generate_model import raw_schema_to_semantic_context
+from semantic_model_generator.protos import semantic_model_pb2
+from semantic_model_generator.protos.semantic_model_pb2 import Dimension, Table
+
+
+SNOWFLAKE_ACCOUNT = os.environ["SNOWFLAKE_ACCOUNT_LOCATOR"]
+_TMP_FILE_NAME = f"admin_app_temp_model_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
+
+
+def update_last_validated_model() -> None:
+    """Whenever user validated, update the last_validated_model to track semantic_model,
+    except for verified_queries field."""
+    st.session_state.last_validated_model.CopyFrom(st.session_state.semantic_model)
+    # Do not save verfieid_queries field for the latest validated.
+    del st.session_state.last_validated_model.verified_queries[:]
+
+
+def changed_from_last_validated_model() -> bool:
+    """Compare the last validated model against latest semantic model,
+    except for verified_queries field."""
+
+    for field in st.session_state.semantic_model.DESCRIPTOR.fields:
+        if field.name != "verified_queries":
+            model_value = getattr(st.session_state.semantic_model, field.name)
+            last_validated_value = getattr(
+                st.session_state.last_validated_model, field.name
+            )
+            if model_value != last_validated_value:
+                return True
+    return False
+
+
+# TODO:
```

Comment:
can you add a REAMDE under `admin_app` with this information + a brief description of the app

Reply:
I added a block of instructions under the about.py st.markdown(), Can you take another look?

Reply:
I still think readme is a more standard place to look